server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -0,0 +1,121 @@
+
+case class PerRecipientRateLimiter(rateLimiter: RateLimiter, keyPrefix: Option[KeyPrefix], entityType: EntityType) {
+  def rateLimit(recipient: MailAddress, mail: Mail): Publisher[RateLimitingResult] =
+    entityType.extractQuantity(mail)
+      .map(increment => rateLimiter.rateLimit(RecipientKey(keyPrefix, entityType, recipient), increment))
+      .getOrElse(SMono.just[RateLimitingResult](RateExceeded))
+}
+
+case class RecipientKey(keyPrefix: Option[KeyPrefix], entityType: EntityType, mailAddress: MailAddress) extends RateLimitingKey {
+  override def asString(): String = s"${
+    keyPrefix.map(prefix => prefix.value + "_")
+      .getOrElse("")
+  }${entityType.asString()}_${mailAddress.asString()}"
+}
+
+class PerRecipientRateLimitMailet @Inject()(rateLimiterFactoryProvider: RateLimiterFactoryProvider) extends GenericMailet {
+  private var exceededProcessor: String = _
+  private var rateLimiters: Seq[PerRecipientRateLimiter] = _
+
+  override def init(): Unit = {
+    val rateLimiterFactory: RateLimiterFactory = rateLimiterFactoryProvider.create(getMailetConfig)
+    val duration: Duration = parseDuration()
+    val keyPrefix: Option[KeyPrefix] = Option(getInitParameter("keyPrefix")).map(KeyPrefix)
+    exceededProcessor = getInitParameter("exceededProcessor", Mail.ERROR)
+
+    val countRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("count"))
+      .flatMap(parameter => extractToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Count)))
+
+    val sizeRateLimiter: Option[PerRecipientRateLimiter] = Option(getInitParameter("size"))
+      .map {
+        case "" => throw new IllegalArgumentException("'size' field cannot be empty if specified")
+        case s => s
+      }
+      .flatMap(parameter => extractToRateLimiter(parameter, duration, rateLimiterFactory)
+        .map(rateLimiter => PerRecipientRateLimiter(rateLimiter, keyPrefix, Size)))
+
+    rateLimiters = Seq(countRateLimiter, sizeRateLimiter)
+      .filter(limiter => limiter.isDefined)
+      .map(limiter => limiter.get)
+  }
+
+  override def service(mail: Mail): Unit = {
+    if (!mail.getRecipients.isEmpty) {
+      val mailAddressAndRateLimitResults: Seq[(MailAddress, RateLimitingResult)] =
+        mail.getRecipients.asScala.map(recipient => (recipient, applyRateLimiter(mail, recipient))).toSeq
+
+      if (allResultsAreExceeded(mailAddressAndRateLimitResults.map(_._2))) {
+        mail.setState(exceededProcessor)
+      } else {
+        mailAddressAndRateLimitResults.foreach {
+          case (mailAddress: MailAddress, rateLimitingResult: RateLimitingResult) =>
+            if (rateLimitingResult.equals(AcceptableRate)) {
+              val newMail: MailImpl = MailImpl.duplicate(mail)
+              newMail.setRecipients(ImmutableList.of(mailAddress))
+              getMailetContext.sendMail(newMail)
+            }
+        }
+      }
```

Comment:
- 1. We should send one email for all the recipients for whose rate is exceeded IMO
 - 2. We should dispose the duplicated email once it is sent
 - 3. The logic around here is hard to read, IMO we should invest into refactoring the code

Reply:
> We should send one email for all the recipients for whose rate is exceeded IMO

Your mean is to send notification mail with content that looks like "You have exceeded the threshold for receiving mail?"

> We should dispose the duplicated email once it is sent

How can we do it?

Reply:
In java 

```
Mail newMail = mail.dispose();
try {
   // do stuff with newMail
} finally {
  Lifecycle.dispose(newMail);
}
```

With Scala we likely can use the Try monad.

Reply:
> Your mean is to send notification mail with content that looks like "You have exceeded the threshold for receiving mail?"

Today if out of 10 recipients I have 5 recipients, you duplicate the email 5 time. Can we duplicate it only one time?

Reply:
> Today if out of 10 recipients I have 5 recipients, you duplicate the email 5 time. Can we duplicate it only one time?

I was clear